base_layer/core/src/transactions/coinbase_builder.rs

```diff
@@ -173,113 +162,167 @@ impl CoinbaseBuilder {
-        let spending_key = self.spend_key.ok_or(CoinbaseBuildError::MissingSpendKey)?;
-        let script_private_key = self.script_key.unwrap_or_else(|| spending_key.clone());
+        let spending_key_id = self.spend_key_id.ok_or(CoinbaseBuildError::MissingSpendKey)?;
+        let script_key_id = self.script_key_id.ok_or(CoinbaseBuildError::MissingScriptKey)?;
+        let covenant = self.covenant;
         let script = self.script.unwrap_or_else(|| script!(Nop));
 
-        let commitment = self
-            .factories
-            .commitment
-            .commit_value(&spending_key, total_reward.as_u64());
-        let output_features = OutputFeatures::create_coinbase(height + constants.coinbase_lock_height(), self.extra);
-        let excess = self.factories.commitment.commit_value(&spending_key, 0);
         let kernel_features = KernelFeatures::create_coinbase();
         let metadata = TransactionMetadata::new_with_features(0.into(), 0, kernel_features);
-        let challenge = TransactionKernel::build_kernel_challenge_from_tx_meta(
-            &TransactionKernelVersion::get_current_version(),
-            &public_nonce,
-            excess.as_public_key(),
-            &metadata,
+        // generate kernel siganture
+        let kernel_version = TransactionKernelVersion::get_current_version();
+        let kernel_message = TransactionKernel::build_kernel_signature_message(
+            &kernel_version,
+            metadata.fee,
+            metadata.lock_height,
+            &metadata.kernel_features,
+            &metadata.burn_commitment,
         );
-        let sig = Signature::sign_raw(&spending_key, nonce, &challenge)
-            .map_err(|_| CoinbaseBuildError::BuildError("Challenge could not be represented as a scalar".into()))?;
-
-        let hasher =
-            DomainSeparatedHasher::<Blake256, WalletServiceHashingDomain>::new_with_label("sender_offset_private_key");
-        let spending_key_hash = hasher.chain(spending_key.as_bytes()).finalize();
-        let spending_key_hash_bytes = spending_key_hash.as_ref();
-
-        let sender_offset_private_key =
-            PrivateKey::from_bytes(spending_key_hash_bytes).map_err(|_| CoinbaseBuildError::InvalidSenderOffsetKey)?;
-        let sender_offset_public_key = PublicKey::from_secret_key(&sender_offset_private_key);
-        let covenant = self.covenant;
-
+        let public_nonce_id = self
+            .key_manager
+            .get_next_key_id(CoreKeyManagerBranch::Nonce.get_branch_key())
+            .await?;
+
+        let public_nonce = self.key_manager.get_public_key_at_key_id(&public_nonce_id).await?;
+        let public_spend_key = self.key_manager.get_public_key_at_key_id(&spending_key_id).await?;
+        let public_script_key = self.key_manager.get_public_key_at_key_id(&script_key_id).await?;
+
+        let kernel_signature = self
+            .key_manager
+            .get_partial_kernel_signature(
```

Comment:
why not just pass in the key_id's here? `public_nonce_id` and `public_nonce` are passed in, when `public_nonce_id` could just be used